base/terminology_processor.bal

```diff
@@ -14,76 +14,1004 @@
+import ballerina/log;
+import ballerina/regex;
+import ballerina/time;
+
+//It is a custom record to hold system URLs and a related concept.  
 type CodeConceptDetails record {
+    //System URL of the concept.
     uri url;
     CodeSystemConcept|ValueSetComposeIncludeConcept concept;
 };
 
-# Function definition for code system finder implementations
-public type CodeSystemFinder isolated function (uri system, code code) returns CodeSystem | ValueSet | FHIRError;
+//It is a custom record to hold system URLs and a array of related concepts.
+type ValueSetExpansionDetails record {
+    //System URL of the concepts.
+    uri url;
+    CodeSystemConcept[]|ValueSetComposeIncludeConcept[] concepts;
+};
 
-# A processor to process terminology data and create relevent data elements
+# The function definition for code system finder implementations.
+public type CodeSystemFinder isolated function (uri system, code code) returns CodeSystem|ValueSet|FHIRError;
+
+# A processor to process terminology data and create relevant data elements.
 public class TerminologyProcessor {
-    
+
+    # Global records to store Terminologies across different profiles and packages.
     private map<CodeSystem> codeSystems = {};
     private map<ValueSet> valueSets = {};
 
     public function init() {
     }
 
-    public function addTerminology(Terminology terminology) {
-        self.addCodeSystems(terminology.codeSystems);
-        self.addValueSets(terminology.valueSets);
+    # It will initialize the Terminology map.
+    # It will delete the existing records if it is re-initialized.
+    # It will be used during the package initialization 
+    #
+    # + terminology - Terminology record
+    public function initTerminology(Terminology terminology) {
+        self.codeSystems = terminology.codeSystems;
+        self.valueSets = terminology.valueSets;
     }
 
-    public function addCodeSystems(CodeSystem[] codeSystems) {
+    # Add list of new CodeSystems to the existing data
+    #
+    # + codeSystems - List CodeSystems
+    # + return - Return List of FHIRErrors if any
+    public function addCodeSystems(CodeSystem[] codeSystems) returns FHIRError[]? {
+        FHIRError[] errors = [];
         foreach CodeSystem codeSystem in codeSystems {
-            self.codeSystems[<string>codeSystem.url] = codeSystem;
+            FHIRError? result = self.addCodeSystem(codeSystem);
+
+            if result is FHIRError {
+                errors.push(result);
+            }
+        }
+
+        return errors;
+    }
+
+    # Add list of new CodeSystems as a json or json array to the existing data
+    #
+    # + codeSystemJson - CodeSystem data in the JSON format
+    # + return - Return List of FHIRErrors if any
+    public function addCodeSystemsAsJson(json[]|json codeSystemJson) returns FHIRError[]? {
+        FHIRError[] errors = [];
+        CodeSystem[] codeSystems = [];
+        json[] codeSystemJsonArray = <json[]>codeSystemJson;
```

Comment:
If we are casting to json[], then we can simply set `codeSystemJson` type as `json[]`. right?

Reply:
Yes, Done